VoodooInput/VoodooInputSimulator/VoodooInputSimulatorDevice.cpp

```diff
@@ -238,53 +231,41 @@ void VoodooInputSimulatorDevice::constructReportGated(const VoodooInputEvent& mu
         input_report.FINGERS[0].Pressure = 0x0;
         input_report.FINGERS[0].Touch_Major = 0x0;
         input_report.FINGERS[0].Touch_Minor = 0x0;
-
         milli_timestamp += 10;
-        
         input_report.timestamp_buffer[0] = (milli_timestamp << 0x3) | 0x4;
         input_report.timestamp_buffer[1] = (milli_timestamp >> 0x5) & 0xFF;
         input_report.timestamp_buffer[2] = (milli_timestamp >> 0xd) & 0xFF;
-        
-        buffer_report = IOBufferMemoryDescriptor::inTaskWithOptions(kernel_task, 0, total_report_len);
-        buffer_report->writeBytes(0, &input_report, total_report_len);
-        handleReport(buffer_report, kIOHIDReportTypeInput);
-        buffer_report->release();
-        
+        handleReport(kernel_buffer, kIOHIDReportTypeInput);
+
         input_report.FINGERS[0].Priority = 0x5;
         input_report.FINGERS[0].State = 0x0;
-        
-        buffer_report = IOBufferMemoryDescriptor::inTaskWithOptions(kernel_task, 0, total_report_len);
-        buffer_report->writeBytes(0, &input_report, total_report_len);
-        handleReport(buffer_report, kIOHIDReportTypeInput);
-        buffer_report->release();
-        
+        handleReport(kernel_buffer, kIOHIDReportTypeInput);
+
         milli_timestamp += 10;
-        
         input_report.timestamp_buffer[0] = (milli_timestamp << 0x3) | 0x4;
         input_report.timestamp_buffer[1] = (milli_timestamp >> 0x5) & 0xFF;
         input_report.timestamp_buffer[2] = (milli_timestamp >> 0xd) & 0xFF;
-
-        buffer_report = IOBufferMemoryDescriptor::inTaskWithOptions(kernel_task, 0, 12);
-        buffer_report->writeBytes(0, &input_report, 12);
-        handleReport(buffer_report, kIOHIDReportTypeInput);
-        buffer_report->release();
+        handleReport(kernel_buffer, kIOHIDReportTypeInput);
     }
-    
     input_report = {};
 }
 
 bool VoodooInputSimulatorDevice::start(IOService* provider) {
     if (!super::start(provider))
         return false;
-    
     ready_for_reports = false;
-    
+
+    kernel_buffer = IOMemoryDescriptor::withAddressRange((mach_vm_address_t)&input_report, sizeof(input_report), kIODirectionNone, kernel_task);
+    kernel_buffer->prepare();
```

Comment:
It is necessary to check for NULL after assigning `kernel_buffer` a value, before executing `kernel_buffer->prepare();`.  That's also the cause of the memory leak below.